my_enumerables.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+# rubocop : disable Metrics/ModuleLength
+# rubocop : disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+# rubocop : disable Metrics/MethodLength
+# rubocop : disable Style/Documentation
+# rubocop : disable Metrics/AbcSize
+
+module Enumerable
+  def my_each
+    if is_a?(Array)
+      0.upto size - 1 do |val|
+        yield(self[val]) if block_given?
+      end
+    elsif is_a?(Range)
+      each do |val|
+        yield(val) if block_given?
+      end
+    else
+      hash_keys = keys
+      0.upto hash_keys.length - 1 do |num|
+        key = hash_keys[num]
+        yield(key, self[key])
+      end
+    end
+  end
+
+  def my_each_with_index
+    0.upto size - 1 do |val|
+      yield(self[val], val)
+    end
+  end
```

Comment:
* [ ] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_each_with_index returns an enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each_with_index.class # not Enumerator
  ```

* [ ] my_each_with_index when a block is given after calling the given block once for each element in self, returns the array itself
   ```ruby
    array = [1,2,5,8,7,98,4]
    block = proc { |num| num < 8 }
    p array.my_each_with_index(&block) == array.each_with_index(&block) # false
  ```